From 341fedaeb1fa35529f59914ba1901ca9d60d0c80 Mon Sep 17 00:00:00 2001
From: Xiaoguang Chen <xiaoguang.chen@intel.com>
Date: Fri, 10 Jul 2015 00:41:55 +0800
Subject: [PATCH 116/403] kvmgt porting MPT support

This patch include kvm specific implementation of mediated
pass-through framework.

Signed-off-by: Jike Song <jike.song@intel.com>
Signed-off-by: Xiaoguang Chen <xiaoguang.chen@intel.com>
diff --git a/drivers/gpu/drm/i915/vgt/vgt.c b/drivers/gpu/drm/i915/vgt/vgt.c
index 3a92823..1e808f2 100644
--- a/drivers/gpu/drm/i915/vgt/vgt.c
+++ b/drivers/gpu/drm/i915/vgt/vgt.c
@@ -36,6 +36,7 @@ MODULE_LICENSE("GPL and additional rights");
 MODULE_VERSION("0.1");
 
 extern struct kernel_dm xengt_kdm;
+extern struct kernel_dm kvmgt_kdm;
 struct kernel_dm *vgt_pkdm = NULL;
 
 bool hvm_render_owner = false;
@@ -1244,10 +1245,14 @@ bool i915_start_vgt(struct pci_dev *pdev)
 	vgt_ops = &__vgt_ops;
 
 	vgt_pkdm = try_then_request_module(symbol_get(xengt_kdm), "xengt");
-	if (vgt_pkdm == NULL) {
+	if (vgt_pkdm == NULL || !hypervisor_check_host()) {
 		printk("vgt: Could not load xengt MPT service\n");
-		return false;
-	} //TODO: request kvmgt here!
+		vgt_pkdm = try_then_request_module(symbol_get(kvmgt_kdm), "kvm");
+		if (vgt_pkdm == NULL) {
+			vgt_warn("vgt: Could not load kvmgt MPT service\n");
+			return false;
+		}
+	}
 
 	if (!vgt_check_host())
 		return false;
